rfcs/0000-atomic-volatile.md

```diff
@@ -0,0 +1,430 @@
+compromise of `load_not_atomic()` and `store_not_atomic()` is still kept around,
+only discouraging its use.
+
+---
+
+It should be noted that switching to LLVM's atomic volatile accesses does not
+resolve the second problem very well per se, because although oversized volatile
+accesses will not compile anymore, they will only be "reported" to the user via
+LLVM crashes. This is not a nice user experience, which is why volatile wrapper
+types are proposed.
+
+Their design largely mirrors that of Rust's existing atomic types, which is only
+appropriate since they do expose atomic operations. One goal of this design is
+that it should be possible to re-use the architecture-specific code that already
+exists to selectively expose Rust atomic types and operations depending on what
+the hardware supports under the hood.
+
+---
+
+This RFC currently proposes to expose the full power of LLVM's atomic volatile
+operations, including e.g. read-modify-write operations like compare-and-swap,
+because there are legitimately useful use cases for these operations in
+interprocess communication scenarios.
+
+However, the fact that these operations do not necessarily compile into a single
+hardware instruction is arguably a footgun for volatile's use cases, and it
+could be argued that initially only stabilizing loads, stores and `Relaxed`
+atomic ordering would be more prudent. I'll go back to this in the alternatives
+section of this RFC.
+
+---
+
+As currently designed, this RFC uses `arbitrary_safe_types` to give method-like
+semantics to a `NonNull` raw pointer. I believe that this is necessary to get
+reasonable ergonomics with an atomics-like wrapper type approach. However, it
+could also be argued that a `VolatileXyz::store(ptr, data, ordering)` style of
+API could reasonably work, and avoid coupling with unstable features. Similarly,
+the use of `NonNull` itself could be debated. I'll come back to these points in
+the alternatives section as well.
+
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+Deprecating the existing volatile operations will cause language churn. Not
+deprecating them will keep around two different and subtly incompatible ways to
+do the same thing, which is equally bad. It could be argued that the issues with
+the existing volatile operations, while real, do not warrant the full complexity
+of this RFC.
+
+Atomic volatile operations are also a somewhat LLVM-specific concept, and
+requiring them may make the life of our other compiler backends harder.
+
+As mentioned above, exposing more than loads and stores, and non-`Relaxed`
+atomic memory orderings, also muddies the "a volatile op should compile into one
+hardware memory instruction" narrative that is so convenient for loads and stores.
+Further, compatibility of non-load/store atomics in IPC scenario may require
+some ABI agreement on how atomics should be implemented between the interacting
+programs.
```

Comment:
Can we just support `Relaxed` as an ordering initially ? And / or maybe make the other supported orderings platform-dependent ? (e.g. some platforms might support `Release`, `Acquire`, and `SeqCst`, but not all of them need to ?)

Reply:
We totally can support just `Relaxed` initially, I think I mentioned it in the alternatives section somewhere.

I do not see the point in making ordering support platform-dependent, since @Amanieu told me on Zulip that we do not currently support platforms which support some C11 atomic orderings but not all of them (assuming that such platforms exist).

Reply:
If you feel that this is the direction which this proposal should take, and not just that it should mention that as alternate route, feel free to ping me back about it.